### PR TITLE
ENH: stats: add Somers' D test

### PR DESCRIPTION
**Pull request by mdhaber**
In the [CZI Proposal](https://warrenweckesser.github.io/czi/scipy-2019-czi-proposal-revised.pdf), we indicated that we would add [Somers' _D_ test](https://www.jstor.org/stable/2090408).

This is part of our effort to address the top level [Statistics Enhancements roadmap item](https://docs.scipy.org/doc/scipy/reference/roadmap.html#statistics-enhancements) "Expand the set of hypothesis tests."

To do:
- [x] Add input validation
- [x] Add tests
- [x] Resolve discrepancy between p-values

~~Correct Wikipedia example~~

Questions:
- `kendalltau` returns NaNs for some corner cases (e.g. empty input). Should this raise errors instead? Pending a decision, the behavior is inconsistent between lists and their equivalent contingency tables for some corner cases.

### CODE REVIEW

**Comment by rlucas7** [comment]
I think Warren has a PR open with a table implemented, 

https://github.com/scipy/scipy/pull/11352/files

Not sure if I misunderstand, It Does look like the interfaces are a bit different though?


Sincerely, 

-Lucas Roberts

**Comment by mdhaber** [comment]
As explained in the [SAS FREQ Procedure Manual](https://support.sas.com/documentation/onlinedoc/stat/151/freq.pdf) (page 3098) and [Brown and Benedetti](https://www.jstor.org/stable/pdf/2286793.pdf), there is a close relationship among Goodman and Kruskal's Gamma, Kendall's tau-b, Stuart's tau-c, and Somers' D. They all have to do with the difference between the number of concordant and discordant pairs in the rankings, and they differ only in how they are normalized to lie between -1 and +1. These normalizations cancel out in the ratio between the statistic and their asymptotic standard error under the null hypothesis (that the true value of the statistic for the population is zero), so the tests for all of the measures are identical. [Somers](https://www.jstor.org/stable/2090408) explains:
> It was pointed out above that D(X|Y) and D(Y|X), like tau-b and gamma, employ Kendall's S in the numerator. Clearly, if the hypothesis that S=0 in the population is accepted, then the hypotheses that D(X|Y)=D(Y|X)=tau_b=gamma=0 are also accepted. Consequently, a test for the significance of Kendall's S is equivalent to a test for the significance of any one of the coefficients, so long as the null hypothesis is that the population value of the coefficient is zero.

Perhaps it is surprising that I believe this implementation to be correct when the *p*-value differs from that of `kendalltau`, which we have reviewed extensively.
```
>>> from scipy.stats import kendalltau, somersd
>>> x = [5, 2, 1, 3, 6, 4, 7]
>>> y = [5, 2, 6, 3, 1, 7, 4]
>>> res = kendalltau(x, y, method='asymptotic')
>>> print(res.correlation, res.pvalue)
-0.14285714285714288 0.6523041372117631
>>> res = somersd(x, y)
>>> print(res.statistic, res.pvalue)  # same statistic in the absence of ties
-0.14285714285714285 0.6303269531576785
```

Further evidence against my implementation comes from the example on [Wikipedia](https://en.wikipedia.org/wiki/Somers%27_D):
```
>>> table = np.array([[3, 5, 2], [1, 7, 6]]).T  # X, the independent variable, should be the row variable
>>> res = somersd(table)
>>> res.statistic  # Wikipedia says 0.34
0.2727272727272727
```

First, the Wikipedia article seems to contradict itself. It states that *X* is independent and *Y* is dependent, yet computes D_XY
image

which would be inconsistent with the original notation.
image

SAS agrees that with X as the independent variable and Y as the independent variable, the statistic is 0.2727; in the reverse situation, the statistic would be 0.3429. This `somersd` implementation also agrees with the value of the statistic in [this SPSS example](https://statistics.laerd.com/spss-tutorials/somers-d-using-spss-statistics.php), so I think that the statistic is calculated correctly.

This brings us back to the *p*-value discrepancy. Surprisingly, although the *p*-value returned by `scipy.stats.kendalltau` agrees with R and SAS PROC CORR:
```
# SciPy
>>> kendalltau([1, 1, 2], [1, 2, 3])
KendalltauResult(correlation=0.816496580927726, pvalue=0.22067136191984682)
```
```
# R
>>> x <- c(1,1,2)
>>> y <- c(1,2,3)
>>> cor.test(x, y, method = "kendall", alternative = "two.sided")

	Kendall's rank correlation tau

data:  x and y
z = 1.2247, p-value = 0.2207
alternative hypothesis: true tau is not equal to 0
sample estimates:
      tau 
0.8164966 
```

The *p* value returned by this `somersd` agrees with SAS PROC FREQ:
```
>>> res = somersd([1, 1, 2], [1, 2, 3])
>>> res.pvalue
0.014305878435429657
```
and SAS PROC FREQ does not agree with SAS PROC CORR.

I haven't yet identified the difference in the assumptions (or approximations) used to compute the *p*-values in these two routines. If you know, please tell!

**Comment by mdhaber** [comment]
@rlucas7 thanks. we've talked about it. We'll consolidate when one of these gets merged.

File: scipy/stats/_hypotests.py [hunk 1]
```diff
@@ -146,3 +146,286 @@ def _get_wilcoxon_distr(n):
...
+class SomersDResult:
+    """
+    A placeholder until we figure out how new stats results should be returned
+    """
+    def __init__(self, statistic, pvalue, table):
+        self.statistic, self.pvalue, self.table, = statistic, pvalue, table
+
+
+def somersd(x, y=None):
+    """
```
Comment (mdhaber) [inline scipy/stats/_hypotests.py:284]:
```suggestion
    r"""
```

File: scipy/stats/_hypotests.py [hunk 2]
```diff
@@ -146,3 +146,286 @@ def _get_wilcoxon_distr(n):
...
+    strong disagreement. They differ in how they are normalized. To show the
+    relationship, Somers' :math:`D` can be defined in terms of Kendall's
+    :math:`\tau_a`:
+
+    .. math::
+        D(Y|X) = \frac{\tau_a(X, Y)}{\tau_a(X, X)}
+
+    Suppose the first ranking :math:`X` has :math:`r` distinct ranks and the
+    second ranking :math:`Y` has :math:`s` distinct ranks. These two lists of
+    :math:`n` rankings can also be viewed as an :math:`r \\times s` contingency
```
Comment (mdhaber) [inline scipy/stats/_hypotests.py:301]:
```suggestion
    :math:`n` rankings can also be viewed as an :math:`r \times s` contingency
```

File: scipy/stats/_hypotests.py [hunk 3]
```diff
@@ -146,3 +146,286 @@ def _get_wilcoxon_distr(n):
...
+    "below average (1)" and a corresponding satisfaction of "very
+    dissatisfied (1)". We perform the analysis as follows.
+
+>>> from scipy.stats import somersd
+>>> table = [[27, 25, 14, 7, 0], [7, 14, 18, 35, 12], [1, 3, 2, 7, 17]]
+>>> res = somersd(table)
+>>> res.statistic
+    0.6032766111513396
+>>> res.pvalue
+    1.0007091191074533e-27
```
Comment (mdhaber) [inline scipy/stats/_hypotests.py:411]:
```suggestion
    >>> from scipy.stats import somersd
    >>> table = [[27, 25, 14, 7, 0], [7, 14, 18, 35, 12], [1, 3, 2, 7, 17]]
    >>> res = somersd(table)
    >>> res.statistic
    0.6032766111513396
    >>> res.pvalue
    1.0007091191074533e-27
```

**Comment by mdhaber** [comment]
To better understand the difference between the *p*-values returned by `somersd` vs `kendalltau`, I performed a computational experiment similar to that in [Brown and Benedetti](https://www.jstor.org/stable/pdf/2286793.pdf). I'm going to write about it from an intuitive perspective rather than one that is mathematically precise.  For example, let's pretend that populations are large but finite.

### Background
Consider a large population of pairs of rankings (X, Y). The null hypothesis used in `kendalltau` is that of independence between X and Y: the population of pairs contains equal numbers of each possible pair.  If we were to perform `kendalltau` (or `somersd`) on the entire population, the statistic would be zero. But if we randomly sample a number of pairs _N_ from the population, the statistic calculated for these _N_ pairs will not necessarily be zero. Instead, there is a distribution of the test statistic under this null hypothesis. To get our *p*-value in `kendalltau`, we compare the statistic of the user-provided data against this distribution. The (two-sided) *p*-value given by `kendalltau` is the percentage of statistics in the null hypothesis distribution that are more extreme than the statistic of the provided table.

Another way of thinking about this distribution is that it arises from sampling contingency tables from a multinomial distribution with *N* trials and the probability associated with each cell is the same *p*.

| *p* | *p* | *p* | *p* |
| *p* | *p* | *p* | *p* |
| *p* | *p* | *p* | *p* |
| *p* | *p* | *p* | *p* |

But this is not the only population for which the statistic is zero. For example, the statistic is zero for all tables with the structure:

| *a* | *b* | *b* | *a* |
| *c* | *d* | *d* | *c* |
| *c* | *d* | *d* | *c* |
| *a* | *b* | *b* | *a* |

We can instead sample contingency tables using a multinomial distribution with *N* trials and probabilities *a*, *b*, *c*, and *d*, and these are also consistent with the null hypothesis that the statistic of the population is zero.

So `kendalltau`'s *p*-values arise from a restrictive null hypothesis of independence, whereas `somers` *p*-values are from the more general null hypothesis that the statistic of the population is zero.

### The experiment
I sampled 10000 contingency tables using a multinomial distribution with _N_=100 and each of the table probability structures shown here.
image
Note that patterns I and and VI are identical; they both correspond with the hypothesis of independence. The authors used the same probability structure twice to see the variation.

For each table, I used `kendalltau` and `somersd` to compute a *p*-value and counted the number of times the null hypothesis was rejected at a confidence level of 5%. We would expect the number of rejections to by 5% of 10000 = 500. This plot summarizes the observed number of rejections (above) or below this number:

image

As we might expect, `kendalltau` rejected the null hypothesis almost exactly as frequently as we would expect for table structures I and VI (which correspond with the hypothesis of independence). For other table structures, the behavior varies quite a bit - sometimes rejecting the null hypothesis far too frequently (e.g. structure ix) and sometimes not enough (e.g. structure v). Overall, `somersd` is more consistent, if a little liberal (i.e. non-conservative).

I repeated the same test with significance level 1%:
image
And we see the same sort of trend.

### Discussion
I think it's OK if `kendalltau` and `somersd` return different *p*-values as long as we explain why, as there is merit to both null hypotheses. If needed, the user could calculate the test statistic using one function and the *p*-value for the null hypothesis they're interested in using the other (as the *p*-value is the same for both statistics under the same null hypothesis). In the long run, we probably want to unify all this, though. For example, although classes aren't a huge part of SciPy's API, I imagine a `ContingencyTable` class (or similar) that can calculate any of the related statistics and corresponding *p*-values / confidence intervals / ASEs for various null hypotheses.

**Comment by mdhaber** [comment]
@swallan Can you help me check the results of tests against SAS PROC FREQ with more than four decimal places of precision (ideally 14-16)? Specifically:

- all the tests in `test_like_kendalltau` that have a hard-coded `expected` numbers
- the tests in `test_asymmetry`
- `test_contingency_table_with_zero_rows_cols`

I've already included numbers for these tests, but they may need to be replaced. In some cases they're tests you already did in SAS, but it only reported 4 decimal places - I'd like higher precision. In other cases, I've just included the current output of `somersd` just so I can get the tests running; obviously these need to be checked. 

There are a few other tests in `test_like_kendalltau` that expect `np.nan`. I'd like to know what SAS does in these cases, but you don't need to change the tests right now. Just comment on what SAS does.

You can either check out my branch and submit a PR on my repo, or you can use GitHub's web interface to suggest the changes.  

Thanks!

File: scipy/stats/_hypotests.py [hunk 4]
```diff
@@ -146,3 +146,334 @@ def _get_wilcoxon_distr(n):
...
+    for i, row in enumerate(table):
+        for j, val in enumerate(row):
+            x[k:k+val] = i + 1
+            y[k:k+val] = j + 1
+            k += val
+    return x, y
+
+
+def _lists_to_ct(x, y):
+    """Convert lists of rankings to contingency table"""
```
Comment (mdhaber) [inline scipy/stats/_hypotests.py:167]: I seem to remember there being a mistake in here. In any case, I think there is existing code for creating a contingency table that I should use.

File: scipy/stats/tests/test_hypotests.py [hunk 5]
```diff
@@ -159,3 +160,253 @@ def test_callable_cdf(self):
         r1 = cramervonmises(x, distributions.beta.cdf, args)
         r2 = cramervonmises(x, "beta", args)
         assert_equal((r1.statistic, r1.pvalue), (r2.statistic, r2.pvalue))
+
+
+def assert_raises_with_match(exception_type, match, function, *args, **kwargs):
+    with assert_raises(exception_type, match=match):
+        function(*args, **kwargs)
```
Comment (WarrenWeckesser) [inline scipy/stats/tests/test_hypotests.py:167]: Instead of creating this function, let's just use the context manager in the tests.  I think the context manager makes a nicer looking test.

File: scipy/stats/tests/test_hypotests.py [hunk 6]
```diff
@@ -159,3 +160,253 @@ def test_callable_cdf(self):
...
+
+
+class TestSomersD(object):
+
+    def test_like_kendalltau(self):
+        # All tests correspond with one in test_stats.py `test_kendalltau
```
Comment (WarrenWeckesser) [inline scipy/stats/tests/test_hypotests.py:173]:
```suggestion
        # All tests correspond with one in test_stats.py `test_kendalltau`
```

File: scipy/stats/tests/test_hypotests.py [hunk 7]
```diff
@@ -159,3 +160,253 @@ def test_callable_cdf(self):
...
+
+
+def assert_raises_with_match(exception_type, match, function, *args, **kwargs):
+    with assert_raises(exception_type, match=match):
+        function(*args, **kwargs)
+
+
+class TestSomersD(object):
+
+    def test_like_kendalltau(self):
```
Comment (WarrenWeckesser) [inline scipy/stats/tests/test_hypotests.py:172]: Not necessary to change now, but this is actually a large collection of tests.  It would be nice to split it into separate, focused tests.  (We don't have to copy the style of `test_kendalltau` *exactly*.)  But let's leave that for a future maintenance PR.

File: scipy/stats/tests/test_hypotests.py [hunk 8]
```diff
@@ -159,3 +160,253 @@ def test_callable_cdf(self):
...
+        assert_allclose(res.statistic, expected[0], atol=1e-15)
+        assert_allclose(res.pvalue, expected[1], atol=1e-15)
+
+        # with only ties in one or both inputs
+        # SAS will not produce an output for these:
+        '''
+        NOTE: No statistics are computed for x * y because x has fewer than 2 nonmissing levels.
+        WARNING: No OUTPUT data set is produced for this table because a row or column variable has fewer
+            than 2 nonmissing levels and no statistics are computed.
+        '''
```
Comment (WarrenWeckesser) [inline scipy/stats/tests/test_hypotests.py:255]: Why is this string included as a statement?  Shouldn't this be a continuation of the comment?
```suggestion
        # With only ties in one or both inputs, SAS will not produce
        # an output for these cases.  The SAS message is:
        # 
        # NOTE: No statistics are computed for x * y because x has fewer
        # than 2 nonmissing levels.
        # WARNING: No OUTPUT data set is produced for this table because a
        # row or column variable has fewer than 2 nonmissing levels and no
        # statistics are computed.
        #
```

**Review by WarrenWeckesser — CHANGES_REQUESTED** [review]
@mdhaber, this is great, and the tests are quite thorough.

My suggested changes are mostly cosmetic, although I would like to get rid of the function `assert_raises_with_match` and just use the context manager in the tests.  I think the context manager makes the tests easier to read.

File: scipy/stats/tests/test_hypotests.py (hunk 8 above)
Comment (mdhaber) [reply scipy/stats/tests/test_hypotests.py:255]: I'm not sure. I don't remember writing that. @swallan did you add that?
Then again, when I saw the `skewcauchy` had issues yesterday, I didn't really remember writing `skewcauchy`, either.

**Comment by mdhaber** [comment]
@WarrenWeckesser tag you're it

**Comment by WarrenWeckesser** [comment]
Thanks @mdhaber, looks good.  If there are no further comments by then, I'll merge this Friday afternoon.

**Review by WarrenWeckesser — APPROVED** [review]

**Comment by WarrenWeckesser** [comment]
Merged, and a brief note added to the [1.7.0 release notes](https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.7.0).  Thanks @mdhaber and @swallan!
